pkg/operator/certrotation/metadata.go

```diff
@@ -0,0 +1,22 @@
+package certrotation
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func ensureMetadataUpdate(secret *corev1.Secret, owner *metav1.OwnerReference, additionalAnnotations AdditionalAnnotations) bool {
+	needsMetadataUpdate := false
+	// no ownerReference set
+	if owner != nil {
+		needsMetadataUpdate = ensureOwnerReference(&secret.ObjectMeta, owner)
+	}
+	// ownership annotations not set
+	needsMetadataUpdate = additionalAnnotations.EnsureTLSMetadataUpdate(&secret.ObjectMeta) || needsMetadataUpdate
+	// convert outdated secret type (created by pre 4.7 installer)
+	if secret.Type != corev1.SecretTypeTLS {
+		secret.Type = corev1.SecretTypeTLS
+		needsMetadataUpdate = true
```

Comment:
Would it be worthwhile to check that the secret we're converting has `tls.crt` and `tls.key` before updating the type, since updating the type makes validation stricter (https://github.com/kubernetes/kubernetes/blob/d51e2da869350b2b20a33d060e17bd1d2165e02d/pkg/apis/core/validation/validation.go#L6392-L6398), or will it be enough to see UpdateSecretFailed event spam?

Reply:
Recreated by whom? We don't react to events but the secret content, right?

Reply:
Recreated by `ApplySecret`, which attempts apply first, but if it fails it does delete+create. If create with `kubernetes.io/tls` and existing content fails, it would throw an error, delete the content and generate a new cert.

Reply:
Updated `ensureMetadataUpdate` to update secret type (removing data if `tls.crt`/`tls.key` are missing) in this case, added unit tests